File: algorithms/CPlusPlus/Recursion/string-length.cpp
/*
Description: find length of string using recursion.

Approach: To run the recursive call till the end of the string.
Maintaining a iterator variable which will increment at every recursive call
If we reach at the end of the string, return the value of the iterator variable
which will be the length of the string.

Time Complexity: O(n) where n is the length of the string
*/
#include <iostream>
using namespace std;

//function starts
int cal_length(string s, int i = 0)
{
    if (s[i] == '\0')
    {
        return i;
    }
    return cal_length(s, i + 1);
}

//main starts
int main()
{
    string s = "hello";
    cout << cal_length(s);
    return 0;
}

/*
Sample Input: hello
Output: 5

Sample Input: good
Output: 4

Sample Input: github
Output: 6
*/

File: algorithms/Java/recursion/string-length.java
package com.dsa;

/*
Description: To find the length of the string using recursion

Time Complexity: O(n) where n is the number of characters present in the string
*/

public class StringLength {

    //function body
    static int findLength(String s) {
        //if we are at the end of the string return 0
        //this is the base case
        if (s.equals("")) {
            return 0;
        }
        //recursive function call
        //at every recursive function call, we are removing the last character of the string using substring() method
        //since we are removing one character at every recursive function call we are adding one to it, to count the character
        return findLength(s.substring(1)) + 1;
    }
    public static void main(String[] args) {
        String s = "Hello";
        //calling the function to return the string length
        int len = findLength(s);
        System.out.println(len);
    }
}
